torch/distributed/_fsdp/utils.py

```diff
@@ -0,0 +1,36 @@
+from typing import Dict, List, Tuple, Union, Any, Callable, Set
+from collections import OrderedDict  # noqa: F401
+
+import torch
+from torch.nn.utils.rnn import PackedSequence
+
+
+"""Useful functions to deal with tensor types with other python container types."""
+
+
+def _apply_to_tensors(fn: Callable, container: Union[torch.Tensor, Dict, List, Tuple, Set]) -> Any:
+    """Recursively apply to all tensor in different kinds of container types."""
+
+    def _apply(x: Union[torch.Tensor, Dict, List, Tuple, Set]) -> Any:
+        if torch.is_tensor(x):
+            return fn(x)
+        elif isinstance(x, OrderedDict):
+            od = OrderedDict()
+            for key, value in x.items():
+                od[key] = _apply(value)
+            return od
+        elif isinstance(x, PackedSequence):
+            _apply(x.data)
+            return x
+        elif isinstance(x, dict):
+            return {key: _apply(value) for key, value in x.items()}
+        elif isinstance(x, list):
```

Comment:
nit: all three last cases below can be done with `type(x)(_apply(el) for el in x)`